haystack/document_stores/pinecone.py

```diff
@@ -435,22 +560,128 @@ def get_all_documents_generator(
 
-        index = index or self.index
-        index = self._sanitize_index_name(index)
-        documents = super(PineconeDocumentStore, self).get_all_documents_generator(
-            index=index, filters=filters, batch_size=batch_size, return_embedding=False
-        )
+        index = str(index or self.index)
+        index = _sanitize_index_name(index)
 
-        for doc in documents:
-            if return_embedding:
-                self._attach_embedding_to_document(document=doc, index=index)
-            yield doc
+        if namespace is None:
+            if self.get_embedding_count() > 0:
+                namespace = self.embedding_namespace
+            else:
+                namespace = self.document_namespace
+
+        ids = self._get_all_document_ids(namespace=namespace, filters=filters, batch_size=batch_size)
+        for i in range(0, len(ids), batch_size):
+            i_end = min(len(ids), i + batch_size)
+            documents = self.get_documents_by_id(
+                ids=ids[i:i_end], namespace=namespace, batch_size=batch_size, return_embedding=return_embedding
+            )
+            for doc in documents:
+                yield doc
+
+    def _get_all_document_ids(
+        self,
+        index: Optional[str] = None,
+        namespace: Optional[str] = None,
+        filters: Optional[Dict[str, Union[Dict, List, str, int, float, bool]]] = None,
+        batch_size: int = 32,
+    ) -> List[str]:
+        index = str(index or self.index)
+        index = _sanitize_index_name(index)
+        if index not in self.pinecone_indexes:
+            raise DocumentStoreError(
+                f"Index named '{index}' does not exist. Try reinitializing PineconeDocumentStore() and running "
+                f"'update_embeddings()' to create and populate an index."
+            )
+
+        if namespace is None:
+            if self.get_embedding_count() > 0:
+                namespace = self.embedding_namespace
+            else:
+                namespace = self.document_namespace
+
+        document_count = self.get_document_count()
+        if len(self.all_ids) == document_count and filters is None:
+            # We have all of the IDs and don't need to extract from Pinecone
+            return list(self.all_ids)
+        else:
+            target_namespace = f"{namespace}-copy"
+            all_ids: Set[str] = set()
+            with tqdm(
+                total=document_count, disable=not self.progress_bar, position=0, unit=" ids", desc="Retrieving IDs"
+            ) as progress_bar:
+                while True:
+                    # Retrieve IDs from Pinecone
+                    vector_id_matrix = self._get_ids(
+                        index=index, namespace=namespace, batch_size=batch_size, filters=filters
+                    )
+                    # Once we reach final item, we break
+                    if len(vector_id_matrix) == 0:
+                        break
```

Comment:
:scroll:  How about using this condition in the `while` loop instead of an internal `break`? Something like:

```python
# Once we reach final item, we leave the loop
while len(vector_id_matrix):     # or `while len(vector_id_matrix) != 0` if that's clearer
    # Retrieve IDs from Pinecone
    vector_id_matrix = self._get_ids(
        index=index, namespace=namespace, batch_size=batch_size, filters=filters
    )
```
You might have to adjust the loop logic, as I see the break occurs in the middle, but it looks doable.

Reply:
Adding w/ next commit